src/scheme.mli

```diff
@@ -3,10 +3,28 @@
 open! Stdune
 
 type 'rules t =
+  (** [Empty] is a scheme that has no rules *)
   | Empty
+
+  (** [Union (a, b)] produces all the rules produced by a and b *)
   | Union of 'rules t * 'rules t
+
+  (** [Approximation (dirs, x)] produces the same rules as [x] in [dirs] and
+      produces no rules outside of [dirs].
+
+      It is an error if [x] produces any rules outside of [dirs].
+      This error is not always going to be detected, especially if it's hidden
+      by an occurrence of [Thunk]. If the error is undetected, the violating
+      rules are just silently ignored. *)
   | Approximation of Dir_set.t * 'rules t
+
+  (** [Finite rules] just produces a fixed set of rules known in advance. *)
   | Finite of 'rules Path.Build.Map.t
+
+  (** [Thunk f] is a "lazy" collection of rules. This is normally used with
+      [Approximation (dirs, Thunk f)] such that the work of [f] can be delayed
+      (or avoided entirely) until (or unless) the rules for [dirs] become
+      necessary. *)
   | Thunk of (unit -> 'rules t)
```

Comment:
Is it guaranteed that the argument of `Thunk` will only be called once?

Reply:
Once per `evaluate`, yes.

Reply:
Okay, I think it's worth clarifying.